File: backend/src/main/java/cn/etarch/mao/harness/delegate/SubagentRecoveryCoordinator.java
package cn.etarch.mao.harness.delegate;

import cn.etarch.mao.harness.core.SessionCrashRecoveryService;
import cn.etarch.mao.harness.delegate.entity.SubagentExecution;
import cn.etarch.mao.harness.delegate.mapper.SubagentExecutionMapper;
import cn.etarch.mao.session.entity.Session;
import cn.etarch.mao.session.entity.SessionCompaction;
import cn.etarch.mao.session.mapper.SessionMapper;
import cn.etarch.mao.session.service.SessionCompactionService;
import cn.etarch.mao.session.service.SessionService;
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class SubagentRecoveryCoordinator {

    private final SubagentExecutionMapper executionMapper;
    private final SessionMapper sessionMapper;
    private final SessionService sessionService;
    private final SessionCompactionService compactionService;
    private final SubagentExecutionRecoveryService recoveryService;
    private final SubagentResultDeliveryService deliveryService;
    private final SessionCrashRecoveryService sessionRecoveryService;
    @Qualifier("agentExecutor")
    private final ExecutorService agentExecutor;

    public void recoverAtStartup() {
        List<SubagentExecution> candidates = executionMapper.selectRecoveryCandidates();
        Map<Long, List<SubagentExecution>> byParent = candidates.stream()
                .collect(Collectors.groupingBy(SubagentExecution::getParentSessionId,
                        LinkedHashMap::new, Collectors.toList()));
        log.info("subagent_recovery_scan executions={} parents={}", candidates.size(), byParent.size());

        Set<Long> blockedParents = byParent.keySet();
        for (Map.Entry<Long, List<SubagentExecution>> entry : byParent.entrySet()) {
            agentExecutor.submit(() -> recoverParentGroup(entry.getKey(), entry.getValue()));
        }

        List<Session> stale = sessionMapper.selectList(new QueryWrapper<Session>()
                .in("phase", "RUNNING", "RESUMING")
                .ne("session_type", "SUBAGENT"));
        for (Session session : stale) {
            if (!blockedParents.contains(session.getId())) {
                agentExecutor.submit(() -> sessionRecoveryService.recover(session));
            }
        }
    }

    void recoverParentGroup(Long parentId, List<SubagentExecution> executions) {
        Session parent = sessionMapper.selectById(parentId);
        if (parent == null || isTerminal(parent.getPhase())) {
            deliveryService.suppressForTerminalParent(parentId);
            return;
        }
        List<Long> ids = executions.stream().map(SubagentExecution::getId).sorted().toList();
        log.info("parent_recovery_wait parent={} childExecutions={}", parentId, ids);
        List<CompletableFuture<Void>> futures = new ArrayList<>();
        for (SubagentExecution execution : executions) {
            if (isActive(execution.getStatus())) {
                futures.add(CompletableFuture.runAsync(() -> {
                    try {
                        recoveryService.recover(execution.getId());
                    } catch (Exception e) {
                        log.error("Subagent recovery escaped executionId={}", execution.getId(), e);
                    }
                }));
            }
        }
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

        parent = sessionMapper.selectById(parentId);
        if (parent == null || isTerminal(parent.getPhase())) {
            deliveryService.suppressForTerminalParent(parentId);
            return;
        }
        SessionCompaction compaction = compactionService.loadValidated(parentId);
        sessionService.cleanupIncompleteTailAfterId(parentId, compactionService.boundaryOf(compaction));
        executions.stream().sorted(Comparator.comparing(SubagentExecution::getId))
                .forEach(execution -> deliveryService.deliver(execution.getId()));

        parent = sessionMapper.selectById(parentId);
        if (parent != null && !isTerminal(parent.getPhase())) {
            log.info("parent_recovery_start parent={}", parentId);
            sessionRecoveryService.recover(parent);
        }
    }

    private boolean isActive(String status) {
        return "RUNNING".equals(status) || "RECOVERING".equals(status);
    }

    private boolean isTerminal(String phase) {
        return "COMPLETED".equals(phase) || "FAILED".equals(phase) || "CANCELLED".equals(phase);
    }
}

File: backend-ts/src/harness/delegate/subagent-recovery-coordinator.ts
import type { SubagentExecution } from '../../session/types.js';
import type { Session, SessionCompactionService, SessionMapper, SessionService } from '../deps.js';
import { harnessLog } from '../log.js';
import type { SubagentExecutionMapper } from './subagent-execution.mapper.js';
import type { SubagentExecutionRecoveryService } from './subagent-execution-recovery.service.js';
import type { SubagentResultDeliveryService } from './subagent-result-delivery.service.js';

export class SubagentRecoveryCoordinator {
  constructor(
    private readonly executionMapper: SubagentExecutionMapper,
    private readonly recoveryService: SubagentExecutionRecoveryService,
    private readonly deliveryService: SubagentResultDeliveryService,
    private readonly sessionMapper: SessionMapper,
    private readonly sessionService: SessionService,
    private readonly compactionService: SessionCompactionService,
    private readonly executor: { submit(fn: () => Promise<void>): void },
  ) {}

  async schedule(recoverParent: (session: Session) => Promise<void>): Promise<Set<number>> {
    const candidates = await this.executionMapper.listRecoveryCandidates();
    const groups = new Map<number, SubagentExecution[]>();
    const blocked = new Set<number>();
    for (const execution of candidates) {
      if (execution.parentSessionId == null) continue;
      const rows = groups.get(execution.parentSessionId) ?? [];
      rows.push(execution);
      groups.set(execution.parentSessionId, rows);
      blocked.add(execution.parentSessionId);
      if (execution.childSessionId != null) blocked.add(execution.childSessionId);
    }
    harnessLog('info', `subagent_recovery_scan executions=${candidates.length} parents=${groups.size}`);
    for (const [parentId, executions] of groups) {
      this.executor.submit(() => this.recoverGroup(parentId, executions, recoverParent));
    }
    return blocked;
  }

  private async recoverGroup(
    parentId: number, executions: SubagentExecution[], recoverParent: (session: Session) => Promise<void>,
  ): Promise<void> {
    harnessLog('info', `parent_recovery_wait parent=${parentId} childExecutions=${executions.map((row) => row.id).join(',')}`);
    let parent = await this.sessionMapper.selectById(parentId);
    if (!parent || isTerminal(parent.phase)) {
      await this.suppressParent(parentId, executions);
      return;
    }
    await Promise.all(executions.map(async (execution) => {
      if (!['RUNNING', 'RECOVERING'].includes(execution.status ?? '')) return;
      try {
        await this.recoveryService.recover(execution);
      } catch (error) {
        harnessLog('error', `Subagent recovery escaped executionId=${execution.id}`, error);
      }
    }));
    parent = await this.sessionMapper.selectById(parentId);
    if (!parent || isTerminal(parent.phase)) {
      await this.suppressParent(parentId, executions);
      return;
    }
    // 先投递真实结果，再补占位。反过来的话，占位 TOOL 会让 deliver 以为配对已完整，
    // 父会话续跑时读到的是「结果丢失」而不是子代理输出。
    for (const execution of [...executions].sort((a, b) => (a.id ?? 0) - (b.id ?? 0))) {
      if (execution.id != null) await this.deliveryService.deliver(execution.id);
    }
    const compaction = await this.compactionService.loadValidated(parentId);
    const boundary = this.compactionService.boundaryOf(compaction);
    await this.sessionService.cleanupIncompleteTailAfterId?.(parentId, boundary);
    parent = await this.sessionMapper.selectById(parentId);
    if (!parent || isTerminal(parent.phase)) {
      await this.suppressParent(parentId, executions);
      return;
    }
    harnessLog('info', `parent_recovery_start parent=${parentId}`);
    await recoverParent(parent);
  }

  /** 父会话已终态：抑制投递，并把仍停在运行中的子会话收成取消。 */
  private async suppressParent(parentId: number, executions: SubagentExecution[]): Promise<void> {
    await this.deliveryService.suppressForParent(parentId);
    for (const execution of executions) {
      const childId = execution.childSessionId;
      if (childId == null) continue;
      const child = await this.sessionMapper.selectById(childId);
      if (!child || isTerminal(child.phase)) continue;
      await this.sessionService.updatePhase(childId, 'CANCELLED');
    }
  }
}

function isTerminal(phase: string | null | undefined): boolean {
  return phase === 'COMPLETED' || phase === 'FAILED' || phase === 'CANCELLED';
}
